carla/recourse_methods/catalog/roar/library/roar.py

```diff
@@ -0,0 +1,196 @@
+
+    delta_opt = res.x
+    delta_W, delta_W0 = np.array(delta_opt[:-1]), np.array([delta_opt[-1]])
+
+    return delta_W, delta_W0
+
+
+def roar_recourse(
+    torch_model,
+    x: np.ndarray,
+    coeff: np.ndarray,
+    intercept: np.float,
+    cat_feature_indices: List[int],
+    binary_cat_features: bool = True,
+    feature_costs: Optional[List[float]] = None,
+    lr: float = 0.01,
+    lambda_param: float = 0.01,
+    delta_max: float = 0.01,
+    y_target: List[int] = [0, 1],
+    t_max_min: float = 0.5,
+    norm: int = 1,
+    loss_type: str = "BCE",
+    loss_threshold: float = 1e-3,
+) -> np.ndarray:
+    """
+    Generates counterfactual example according to ROAR method for input instance x
+
+    Parameters
+    ----------
+    torch_model: black-box-model to discover
+    x: factual to explain
+    coeff: Coefficient for x
+    intercept: Intercept for x
+    cat_feature_indices: list of positions of categorical features in x
+    binary_cat_features: If true, the encoding of x is done by drop_if_binary
+    feature_costs: List with costs per feature
+    lr: learning rate for gradient descent
+    lambda_param: weight factor for feature_cost
+    delta_max: maximum perturbations of weights
+    y_target: Tuple of class probabilities (BCE loss) or [Float] for logit score (MSE loss).
+    t_max_min: maximum time of search
+    norm: L-norm to calculate cost
+    loss_type: String for loss function (MSE or BCE)
+    loss_threshold: Float threshold for loss difference between iterations
+
+    Returns
+    -------
+    Counterfactual example as np.ndarray
+    """
+    device = "cuda" if torch.cuda.is_available() else "cpu"
+    torch.manual_seed(0)
+
+    if feature_costs is not None:
+        feature_costs = torch.from_numpy(feature_costs).float().to(device)
+
+    coeff = torch.from_numpy(coeff).float().to(device)
+    intercept = torch.from_numpy(np.asarray([intercept])).float().to(device)
+    x = torch.from_numpy(x).float().to(device)
+    y_target = torch.tensor(y_target).float().to(device)
+    lamb = torch.tensor(lambda_param).float().to(device)
+
+    # x_new is used for gradient search in optimizing process
+    x_new = Variable(x.clone(), requires_grad=True)
+
+    optimizer = optim.Adam([x_new], lr=lr, amsgrad=True)
+
+    if loss_type == "MSE":
+        if len(y_target) != 1:
+            raise ValueError(f"y_target {y_target} is not a single logit score")
+
+        # If logit is above 0.0 we want class 1, else class 0
+        target_class = torch.tensor(y_target[0] > 0.0).float()
+        loss_fn = torch.nn.MSELoss()
+    elif loss_type == "BCE":
+        if y_target[0] + y_target[1] != 1.0:
+            raise ValueError(
+                f"y_target {y_target} does not contain 2 valid class probabilities"
+            )
+
+        # [0, 1] for class 1, [1, 0] for class 0
+        # target is the class probability of class 1
+        target_class = y_target[1]
```

Comment:
I think this is based on the code in the Wachter method? You might need this:         
`target_class = torch.round(y_target[1]).int()`